### PR TITLE
Add dominator tree

### PR DESCRIPTION
**Pull request by joeleong**
Hi,

This adds dominator trees using the already implemented immediate dominators

Addresses #1941 

This is my first contribution here so let me know if anything needs to be fixed.

Also, out of curiosity what's the consensus opinion on adding very small functions such as explicitly supporting post-dominator trees and immediate post-dominators which are handy, but trivial (too?) to add. In fact, they are already tested with the non-post functions.

Thanks!

### CODE REVIEW

File: networkx/algorithms/dominance.py [hunk 1]
```diff
@@ -135,3 +135,37 @@ def dominance_frontiers(G, start):
...
+    NetworkXError
+        If `start` is not in `G`.
+
+    Notes
+    -----
+    Note that if a CFG has a single exit node,
+    then postdominance is equivalent to dominance if
+    flow is reversed (going from the exit node to the start node)
+    """
+    return nx.DiGraph(((v, u) for u, v in nx.immediate_dominators(G, start).items() if u != start))
```
Comment (jfinkels) [inline networkx/algorithms/dominance.py]: This line is too long (longer than 79 characters). There is also a redundant pair of parentheses. You can split it up into two lines as follows:

```python
    dominators = nx.immediate_dominators(G, start)
    return nx.DiGraph((v, u) for u, v in dominators.items() if u != start)
```

File: networkx/algorithms/dominance.py [hunk 2]
```diff
@@ -135,3 +135,37 @@ def dominance_frontiers(G, start):
...
+    ------
+    NetworkXNotImplemented
+        If `G` is undirected.
+
+    NetworkXError
+        If `start` is not in `G`.
+
+    Notes
+    -----
+    Note that if a CFG has a single exit node,
```
Comment (jfinkels) [inline networkx/algorithms/dominance.py]: Is this note relevant to this particular function? It doesn't seem to mention a dominator tree anywhere. If it is relevant, can you rephrase it so that it applies to any directed graph (since this function is generic)?

File: networkx/algorithms/tests/test_dominance.py [hunk 3]
```diff
@@ -261,3 +261,37 @@ def test_loops_larger(self):
...
+class TestDominatorTree(object):
+
+    def test_dominator_tree(self):
+        # from page 5 (labeled as 388 in bottom right)
+        # http://pages.cs.wisc.edu/~fischer/cs701.f08/lectures/Lecture19.pdf
+        edges = [(1, 2), (2, 3), (2, 4), (3, 5), (4, 5), (5, 6), (6, 7),
+                (7, 6), (7, 8)]
+        G = nx.DiGraph(edges)
+        T = nx.dominator_tree(G, 1)
+        assert_equal(sorted(list(T.edges())),
```
Comment (jfinkels) [inline networkx/algorithms/tests/test_dominance.py]: There is a `networkx.testing.utils.assert_edges_equal` function so that you don't have to sort these things. Something like

```python
from networkx.testing.utils import assert_edges_equal

assert_edges_equal(T.edges(), expected_edges)
```

File: networkx/algorithms/tests/test_dominance.py [hunk 4]
```diff
@@ -261,3 +261,37 @@ def test_loops_larger(self):
...
             assert_equal(set(df[n]),set(answer[n]))
+
+
+class TestDominatorTree(object):
+
+    def test_dominator_tree(self):
+        # from page 5 (labeled as 388 in bottom right)
+        # http://pages.cs.wisc.edu/~fischer/cs701.f08/lectures/Lecture19.pdf
+        edges = [(1, 2), (2, 3), (2, 4), (3, 5), (4, 5), (5, 6), (6, 7),
+                (7, 6), (7, 8)]
```
Comment (jfinkels) [inline networkx/algorithms/tests/test_dominance.py]: The continuation line indentation does not match the line above it, one more space is needed (so the left parentheses line up). You may want to run [`flake8`][1] on your code to check for style issues like this.

[1]: http://flake8.pycqa.org/

File: networkx/algorithms/dominance.py [hunk 5]
```diff
@@ -135,3 +135,37 @@ def dominance_frontiers(G, start):
                         df[v].add(u)
                         v = idom[v]
     return df
+
+
+def dominator_tree(G, start):
+    """Returns the dominator tree from a directed graph.
+
+
```
Comment (jfinkels) [inline networkx/algorithms/dominance.py]: One too many blank lines here, one is enough.

**Comment by MridulS** [comment]
Thanks for this @joeleong!

This hasn't been touched in a while and bunch of things have changed inside networkx in the last 7 years! I'll go ahead and close this PR. Thanks again for your contribution!!
